⚠️ Nightly upstream-dev CI failed ⚠️

[Workflow Run URL](https://github.com/google/jax/actions/runs/6770482126)
<details><summary>Summary of Failures</summary>

```
tests/lax_numpy_test.py::LaxBackedNumpyTests::testCross0: DeprecationWarning: Arrays of 2-dimensional vectors are deprecated. Use arrays of 3-dimensional vectors instead. (deprecated in NumPy 2.0)
tests/scipy_optimize_test.py::TestBFGS::test_minimize0: AssertionError: 
Not equal to tolerance rtol=0.0002, atol=0.0002

Mismatched elements: 2 / 2 (100%)
Max absolute difference among violations: 2.
Max relative difference among violations: 0.6666667
 ACTUAL: array([1., 1.], dtype=float32)
 DESIRED: array([3., 2.], dtype=float32)
```

</details>

Looks like an issue with numpy-nightly:
```
/opt/hostedtoolcache/Python/3.11.5/x64/lib/python3.11/site-packages/ml_dtypes/_finfo.py:19: in <module>
    from ml_dtypes._ml_dtypes_ext import bfloat16
E   ImportError: numpy.core.umath failed to import
------------------------------- Captured stderr --------------------------------
AttributeError: _ARRAY_API not found
AttributeError: _ARRAY_API not found
ImportError: numpy.core._multiarray_umath failed to import
```
I think https://github.com/numpy/numpy/pull/24943 should fix this.

Previous issues fixed; the current failure is this:
```
FAILED tests/scipy_optimize_test.py::TestBFGS::test_minimize0 - AssertionError: 
Not equal to tolerance rtol=0.0002, atol=0.0002

Mismatched elements: 2 / 2 (100%)
Max absolute difference among violations: 2.
Max relative difference among violations: 0.6666667
 ACTUAL: array([1., 1.], dtype=float32)
 DESIRED: array([3., 2.], dtype=float32)
```
I tracked this down to the NEP 50 change in numpy 2.0. The computation in this code has changed:
```python
import numpy as np
def himmelblau(p):
  x, y = p
  return (x ** 2 + y - 11.) ** 2 + (x + y ** 2 - 7.) ** 2
init = np.ones(2, dtype='float32')
print(np.__version__, himmelblau(init).dtype)
```
Results with current numpy:
```
1.26.1 float64
```
Results at head:
```
2.0.0.dev0+git20231025.9f6789c float32
```
and because of this, scipy's optimizer does the computation in 32-bit precision and fails to converge.

--- a/tests/scipy_optimize_test.py
+++ b/tests/scipy_optimize_test.py
@@ -90,7 +90,10 @@ class TestBFGS(jtu.JaxTestCase):
       return result.x
 
     jax_res = min_op(x0)
-    scipy_res = scipy.optimize.minimize(func(np), x0, method='BFGS').x
+    # Newer scipy versions perform poorly in float32. See
+    # https://github.com/scipy/scipy/issues/19024.
+    x0_f64 = x0.astype('float64')
+    scipy_res = scipy.optimize.minimize(func(np), x0_f64, method='BFGS').x
     self.assertAllClose(scipy_res, jax_res, atol=2e-4, rtol=2e-4,
                         check_dtypes=False)
 
@@ -166,10 +169,13 @@ class TestLBFGS(jtu.JaxTestCase):
 
     jax_res = min_op(x0)
 
+    # Newer scipy versions perform poorly in float32. See
+    # https://github.com/scipy/scipy/issues/19024.
+    x0_f64 = x0.astype('float64')
     # Note that without bounds, L-BFGS-B is just L-BFGS
     with jtu.ignore_warning(category=DeprecationWarning,
                             message=".*tostring.*is deprecated.*"):
-      scipy_res = scipy.optimize.minimize(func(np), x0, method='L-BFGS-B').x
+      scipy_res = scipy.optimize.minimize(func(np), x0_f64, method='L-BFGS-B').x
 
     if func.__name__ == 'matyas':
       # scipy performs badly for Matyas, compare to true minimum instead

